smartcontract/service/native/param.go

```diff
@@ -0,0 +1,125 @@
+	cstates "github.com/ontio/ontology/core/states"
+	scommon "github.com/ontio/ontology/core/store/common"
+	ctypes "github.com/ontio/ontology/core/types"
+	"github.com/ontio/ontology/errors"
+	"github.com/ontio/ontology/smartcontract/service/native/states"
+	"sync"
+)
+
+type ParamCache struct {
+	lock   sync.RWMutex
+	Params map[string]string
+}
+
+var GLOBAL_PARAM = map[string]string{
+	"init-key1": "init-value1",
+	"init-key2": "init-value2",
+	"init-key3": "init-value3",
+	"init-key4": "init-value4",
+}
+
+var paramCache *ParamCache
+var admin common.Address
+
+func init() {
+	Contracts[genesis.ParamContractAddress] = RegisterParamContract
+}
+
+func ParamInit(native *NativeService) error {
+	paramCache = new(ParamCache)
+	paramCache.Params = make(map[string]string)
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	for k, v := range GLOBAL_PARAM {
+		native.CloneCache.Add(scommon.ST_STORAGE, getParamKey(contract, k), getParamStorageItem(v))
+		paramCache.Params[k] = v
+	}
+	admin = ctypes.AddressFromPubKey(account.GetBookkeepers()[0])
+	native.CloneCache.Add(scommon.ST_STORAGE, getAdminKey(contract), getAdminStorageItem(admin))
+	return nil
+}
+
+func TransferAdmin(native *NativeService) error {
+	if !native.ContextRef.CheckWitness(admin) {
+		return errors.NewErr("[Transfer Admin]Authentication failed!")
+	}
+	destinationAdmin := new(common.Address)
+	if err := destinationAdmin.Deserialize(bytes.NewBuffer(native.Input)); err != nil {
+		return errors.NewErr("[Transfer Admin]Deserialize Admins failed!")
+	}
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	native.CloneCache.Add(scommon.ST_STORAGE, getAdminKey(contract), getAdminStorageItem(*destinationAdmin))
+	return nil
+}
+
+func ApproveAdmin(native *NativeService) error {
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	destinationAdmin, err := getAdmin(native, getAdminKey(contract))
+	if err != nil {
+		return nil
+	}
+	admin = destinationAdmin
+	log.Warn("[Approve Admin] success, new admin is ", destinationAdmin)
+	return nil
+}
+
+func SetParam(native *NativeService) error {
+	if !native.ContextRef.CheckWitness(admin) {
+		return errors.NewErr("[Set Param]Authentication failed!")
+	}
+	params := new(states.Params)
+	if err := params.Deserialize(bytes.NewBuffer(native.Input)); err != nil {
+		return errors.NewErr("[Set Param]Deserialize failed!")
+	}
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	for _, param := range params.ParamList {
+		setParamToCache(param.K, param.V)
+		native.CloneCache.Add(scommon.ST_STORAGE, getParamKey(contract, param.K), getParamStorageItem(param.V))
+		notifyParamSetSucess(native, contract, param)
+	}
+	return nil
+}
+
+func setParamToCache(key, value string) {
+	paramCache.lock.Lock()
+	defer paramCache.lock.Unlock()
+	paramCache.Params[key] = value
+}
+
+func getParamFromCache(key string) string {
+	paramCache.lock.RLock()
+	defer paramCache.lock.RUnlock()
```

Comment:
if we update node server, this cache will clear. it will can't get params from cache.